### PR TITLE
Introduced a constraint on the Hex tile index.

### PR DESCRIPTION
**Pull request by xavierlepretre**
There are a lot of `Int` thrown around for indices of hex tiles and tile sides. It is difficult to know which is which. The risk is to have run time errors.
So to convert most of them into compile time errors, I propose to introduce basic types that encapsulate and enforce boundaries. This `HexTileIndex` is just the first shot. A demonstration if you will.
Does this sound reasonable?

### CODE REVIEW

**Comment by xavierlepretre** [comment]
Eventually, we also want to make a new type that wraps around `MultableList<HexTileType>` and only permits actions we approve of.

File: contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt [hunk 1]
```diff
@@ -110,8 +119,53 @@ data class AccessPoint(val hexTileIndex: HexTileIndex, val hexTileCoordinate: Li
...
+            else value - 1)
+
+    fun next() = TileSideIndex(
+            if (value + 1 < HexTile.SIDE_COUNT) value + 1
+            else 0)
+}
+
+/**
+ * Applies to settlements.
+ * TODO Confirm that corner 1 is between side 0 and side 1.
```
Comment (xavierlepretre) [inline contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt]: @rogersanick A question for you.

File: contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt [hunk 2]
```diff
@@ -138,17 +161,21 @@ data class TileSideIndex(val value: Int) {
...
     fun previous() = TileSideIndex(
-            if (value - 1 < 0) HexTile.SIDE_COUNT - 1
+            if (value < 1) HexTile.SIDE_COUNT - 1
             else value - 1)
 
     fun next() = TileSideIndex(
             if (value + 1 < HexTile.SIDE_COUNT) value + 1
             else 0)
+
+    // TODO Confirm that side 0 is between corner 0 and corner 1.
```
Comment (xavierlepretre) [inline contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt:334]: @rogersanick That's the updated question.

File: contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt [hunk 3]
```diff
@@ -42,21 +57,26 @@ data class GameBoardState(val hexTiles: MutableList<HexTile>,
 class HexTile(val resourceType: String,
```
Comment (xavierlepretre) [inline contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt]: @rogersanick Any reason why it is not a `data class`?

File: workflows/src/main/kotlin/com/flows/SetupGameBoardFlow.kt [hunk 4]
```diff
@@ -116,11 +116,11 @@ class SetupGameBoardFlow(val p1: Party, val p2: Party, val p3: Party, val p4: Pa
...
-                PortTile(listOf(3 of Wheat, 3 of Wood, 3 of Sheep, 3 of Brick, 3 of Sheep), listOf(1 of Wheat, 1 of Wood, 1 of Sheep, 1 of Brick, 1 of Sheep)),
-                PortTile(listOf(3 of Wheat, 3 of Wood, 3 of Sheep, 3 of Brick, 3 of Sheep), listOf(1 of Wheat, 1 of Wood, 1 of Sheep, 1 of Brick, 1 of Sheep))
+                PortTile(listOf(2 of Wheat), listOf(1 of Wood, 1 of Sheep, 1 of Brick, 1 of Ore)),
+                PortTile(listOf(3 of Wheat, 3 of Wood, 3 of Sheep, 3 of Brick, 3 of Ore), listOf(1 of Wheat, 1 of Wood, 1 of Sheep, 1 of Brick, 1 of Ore)),
+                PortTile(listOf(3 of Wheat, 3 of Wood, 3 of Sheep, 3 of Brick, 3 of Ore), listOf(1 of Wheat, 1 of Wood, 1 of Sheep, 1 of Brick, 1 of Ore)),
+                PortTile(listOf(3 of Wheat, 3 of Wood, 3 of Sheep, 3 of Brick, 3 of Ore), listOf(1 of Wheat, 1 of Wood, 1 of Sheep, 1 of Brick, 1 of Ore)),
+                PortTile(listOf(3 of Wheat, 3 of Wood, 3 of Sheep, 3 of Brick, 3 of Ore), listOf(1 of Wheat, 1 of Wood, 1 of Sheep, 1 of Brick, 1 of Ore))
```
Comment (xavierlepretre) [inline workflows/src/main/kotlin/com/flows/SetupGameBoardFlow.kt:121]: @rogersanick I hope it is a fix indeed.

File: contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt (hunk 1 above)
Comment (rogersanick) [reply contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt]: Hey @xavierlepretre - you are correct.

Corner 0 is the top-most point of a hextile (between sides 5 and 0)
Corner 1 is the point immediately to the right (one jump clockwise and between sides 0 and 1)

File: contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt (hunk 2 above)
Comment (rogersanick) [reply contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt:334]: This is correct! Side 0 is between corner 0 and 1

File: contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt (hunk 3 above)
Comment (rogersanick) [reply contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt]: None imparticular! We can definitely convert it to a data class if required

File: workflows/src/main/kotlin/com/flows/SetupGameBoardFlow.kt (hunk 4 above)
Comment (rogersanick) [reply workflows/src/main/kotlin/com/flows/SetupGameBoardFlow.kt:121]: It is! Thank you very much. This is actually one of the reasons the tests are failing right now.

**Comment by rogersanick** [comment]
Hey @xavierlepretre - this is extremely helpful! I'm keen on merging in your fix for the PortTile prices. When you are ready feel free to mark this as ready for review.

File: workflows/src/main/kotlin/com/flows/SetupGameBoardFlow.kt (hunk 4 above)
Comment (xavierlepretre) [reply workflows/src/main/kotlin/com/flows/SetupGameBoardFlow.kt:121]: Maybe you can just cherry-pick this commit into master. If it is possible.

File: contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt [hunk 5]
```diff
@@ -66,31 +86,144 @@ class HexTile(val resourceType: String,
...
 @CordaSerializable
 data class Port(val portTile: PortTile, var accessPoints: List<AccessPoint>)
 
 @CordaSerializable
-data class AccessPoint(val hexTileIndex: Int, val hexTileCoordinate: List<Int>)
\ No newline at end of file
+data class AccessPoint(val hexTileIndex: HexTileIndex, val hexTileCoordinate: List<TileCornerIndex>)
+
+@CordaSerializable
+data class HexTileNeighbors(private val value: MutableList<HexTileIndex?> = MutableList(HexTile.SIDE_COUNT) { null }) {
```
Comment (xavierlepretre) [inline contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt]: @rogersanick I discovered that when a class is deserialised, the initial value passed as `MutableList` is actually not mutable... This is causing problems as early as `tb.verify(serviceHub)`.
Is there a workaround or should we migrate to `List`?

BTW, the `private val` is a mistake, I am fixing it to `val` in a future commit.

Comment (xavierlepretre) [reply contracts/src/main/kotlin/com/contractsAndStates/states/GameBoardState.kt]: Another way to go would be to have 2 types, the serialisable one with `List`, and the not serialisable one with `MutableList`.
Sounds like the builder pattern.

**Comment by xavierlepretre** [comment]
@rogersanick If you could merge this one. 
There can be some adjustments with the flows so that they make full use of the new structures, but this can be done in another PR.

**Pull request merged**